content/300-guides/100-prisma-guides/100-prisma-client-transactions-guide.mdx

```diff
@@ -0,0 +1,773 @@
+  name  String @unique
+  seats Seat[]
+}
+```
+
+The following sample code finds the first available seat and assigns that seat to a user:s
+
+```ts
+const movieName = "Hidden Figures";
+
+// Find first available seat
+const availableSeat = await client.seat.findFirst({
+  where: {
+    movie: {
+      name: movieName,
+    },
+    claimedBy: null,
+  },
+})
+
+// Throw an error if no seats are available
+if (!availableSeat) {
+  throw new Error(`Oh no! ${movieName} is all booked.`)
+}
+
+// Claim the seat
+await client.seat.update({
+  data: {
+    claimedBy: userId,
+  },
+  where: {
+    id: availableSeat.id,
+  },
+})
+```
+
+However, this code suppers from the "double booking problem" - it is possible for two people to book the same seats:
+
+1. Seat 3A returned to Sorcha (`findFirst`)
+2. Seat 3A returned to Ellen (`findFirst`)
+3. Seat 3A claimed by Sorcha (`update`)
+4. Seat 3A claimed by Ellen (`update` - overwrites Sorcha's claim)
+
+Even though Sorcha has successfully booked the seat, the system ultimately stores Ellen's claim. To solve this problem with optimistic concurrency control, add a `version` field to the seat:
+
+```prisma highlight=7;normal
+model Seat {
+  id        Int   @id @default(autoincrement())
+  userId    Int?
+  claimedBy User? @relation(fields: [userId], references: [id])
+  movieId   Int
+  Movie     Movie @relation(fields: [movieId], references: [id])
+  version   Int
+}
+```
+
+Next, adjust the code to check the `version` field before updating:
+
+```ts highlight=18-37;normal
+const movieName = "Hidden Figures";
+
+// Find the first available seat
+// availableSeat.version might be 0
+const availableSeat = await client.seat.findFirst({
+  where: {
+    Movie: {
+      name: movieName,
+    },
+    claimedBy: null,
+  },
+})
+
+if (!availableSeat) {
+  throw new Error(`Oh no! ${movieName} is all booked.`)
+}
+
+// Only mark the seat as claimed if the availableSeat.version 
+// matches the version we're updating. Additionally, increment the
+// version when we perform this update so all other clients trying
+// to book this same seat will have an outdated version.
+const seats = await client.seat.updateMany({
+  data: {
+    claimedBy: userId,
+    version: {
+      increment: 1,
+    },
+  },
+  where: {
+    id: availableSeat.id,
+    version: availableSeat.version, // This version field is the key; only claim seat if in-memory version matches database version, indicating that the field has not been updated
+  },
+})
+
+if (seats.count === 0) {
+  throw new Error(`That seat is already booked! Please try again.`)
+}
+```
+
+It is now impossible for two people to book the same seat:
+
+1. Seat 3A returned to Sorcha (`version` is 0)
+2. Seat 3A returned to Ellen (`version` is 0)
+3. Seat 3A claimed by Sorcha (`version` is incremented to 1, booking succeeds)
+4. Seat 3A claimed by Ellen (in-memory `version` (0) does not match database `version` (1) - booking does not succeed)
```

Comment:
Some kind of conclusion would be nice.
I would also add why OCC is useful: the fact that it avoids locking makes the application much more resilient to a high number of concurrent requests and hence is more scalable overall.
Locking may end up impacting the performance of the application, or, in the worst-case scenario, lead to an outage in case the application receives a lot of concurrent requests.